cc/transport/grpc_streaming_transport.h

```diff
@@ -17,17 +17,100 @@
  public:
-  absl::StatusOr<::oak::session::v1::AttestationBundle> GetEvidence() override;
-  absl::StatusOr<std::string> Invoke(absl::string_view request_bytes) override;
+  explicit GrpcStreamingTransport(
+      std::unique_ptr<::grpc::ClientReaderWriter<::oak::session::v1::RequestWrapper,
+                                                 ::oak::session::v1::ResponseWrapper>>
+          channel_reader_writer)
+      : channel_reader_writer_(std::move(channel_reader_writer)) {}
+
+  absl::StatusOr<::oak::session::v1::AttestationBundle> GetEvidence() override {
+    // Create request.
+    ::oak::session::v1::RequestWrapper request;
+    ::oak::session::v1::GetPublicKeyRequest get_public_key_request;
+    *request.mutable_get_public_key_request() = get_public_key_request;
+
+    // Send request.
+    auto response = Send(request);
+    if (!response.ok()) {
+      return response.status();
+    }
+
+    // Process response.
+    switch (response->response_case()) {
+      case ::oak::session::v1::ResponseWrapper::kGetPublicKeyResponseFieldNumber:
+        return response->get_public_key_response().attestation_bundle();
+      case ::oak::session::v1::ResponseWrapper::kInvokeResponseFieldNumber:
+        return absl::InternalError("received InvokeResponse instead of GetPublicKeyResponse");
+      case ::oak::session::v1::ResponseWrapper::RESPONSE_NOT_SET:
+      default:
+        return absl::InternalError("received unsupported response: " + response->DebugString());
+    }
+  }
+
+  absl::StatusOr<std::string> Invoke(absl::string_view request_bytes) override {
+    // Create request.
+    ::oak::session::v1::RequestWrapper request;
+    ::oak::session::v1::InvokeRequest* invoke_request = request.mutable_invoke_request();
+    invoke_request->set_encrypted_body(request_bytes);
+
+    // Send request.
+    auto response = Send(request);
+    if (!response.ok()) {
+      return response.status();
+    }
+
+    // Process response.
+    switch (response->response_case()) {
+      case ::oak::session::v1::ResponseWrapper::kGetPublicKeyResponseFieldNumber:
+        return absl::InternalError("received GetPublicKeyResponse instead of InvokeResponse");
+      case ::oak::session::v1::ResponseWrapper::kInvokeResponseFieldNumber:
+        return response->invoke_response().encrypted_body();
+      case ::oak::session::v1::ResponseWrapper::RESPONSE_NOT_SET:
+      default:
+        return absl::InternalError("received unsupported response: " + response->DebugString());
+    }
+  }
+
+ private:
+  std::unique_ptr<::grpc::ClientReaderWriter<::oak::session::v1::RequestWrapper,
+                                             ::oak::session::v1::ResponseWrapper>>
+      channel_reader_writer_;
+
+  absl::StatusOr<::oak::session::v1::ResponseWrapper> Send(
+      const ::oak::session::v1::RequestWrapper& request) {
+    // Send a request.
+    channel_reader_writer_->Write(request);
+    channel_reader_writer_->WritesDone();
```

Comment:
Calling WritesDone here worries me as I think it may shut down the stream (which iiuc needs to be open for both GetPublicKey and Invoke). https://github.com/grpc/grpc/blob/86d7c8125e8210fd96c0b8b2288bc3c96d7cf390/include/grpcpp/support/sync_stream.h#L273-L280

Should we keep it open until after the Invoke method is called? Please add tests here if you disagree.

Reply:
Will be done in a separate PR

Reply:
IIUC the `WritesDone` only tells the server that the stream of messages has been sent: https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_reader_writer_interface.html#a52f4e5d5ac7fe0e4995cb337aa0ecfc8

Tried testing it without `WritesDone`, and the client just hangs.

Reply:
No, looks like `WritesDone` actually closes the stream.

Added an integration here: https://github.com/project-oak/oak/pull/4176

Reply:
> IIUC the `WritesDone` only tells the server that the stream of messages has been sent: https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_reader_writer_interface.html#a52f4e5d5ac7fe0e4995cb337aa0ecfc8
> 
> Tried testing it without `WritesDone`, and the client just hangs.

I think WritesDone needs to be called to tell the server that the client is done sending messages. This can be done in the deconstructor of the transport class or at the end of Invoke() [assuming we are ok with the stream ending after the client sends a single encrypted request]